core/utils/index_utils.hpp

```diff
@@ -24,6 +24,320 @@
+          auto rMeta = right.meta;
+          auto lLivePerc = static_cast<double>(lMeta->live_docs_count) / lMeta->docs_count;
+          auto rLivePerc = static_cast<double>(rMeta->live_docs_count) / rMeta->docs_count;
+
+          return (lLivePerc < rLivePerc);
+        };
+
+        std::sort(segments.begin(), segments.end(), segmentSortFunc);
+
+        auto count = 0;
+        auto total_docs_count = 0;
+        auto total_live_docs_count = 0;
+        double livePerc;
+
+        for (auto itr = segments.begin(); itr != segments.end(); itr++) {
+          auto meta = itr->meta;
+          total_docs_count += meta->docs_count;
+          total_live_docs_count += meta->live_docs_count;
+
+          livePerc = static_cast<double>(total_live_docs_count) / total_docs_count;
+          if (livePerc > tier::ConsolidationConfig::maxLivePercentage)
+            break;
+
+          ++count;
+        }
+
+        if (count < 1)
+          return false;
+
+        best = ConsolidationCandidate<Segment>(segments.begin(), segments.begin() + count - 1, getSegmentAttributes);
+        return true;
+    }
+
+    //
+    //  This function receives a sorted vector of segments
+    //  and finds the best contiguous subset of segments to
+    //  merge together.
+    //  The best subset is defined as the one with the lowest
+    //  merge cost and the merge cost is computed inside the
+    //  ConslidateCandidate struct.
+    //
+    //  findBestConsolidationCandidate merely sets a rolling
+    //  window of size 4 on a subset of segments, it lets
+    //  ConsolidationCandidate compute the cost of merge for that
+    //  subset, repeats this process for all contiguous subsets
+    //  and identifies the best candidate.
+    //
+    template<typename Segment>
+    bool findBestConsolidationCandidate(
+      std::vector<Segment>& sorted_segments,
+      size_t max_segments_bytes,
+      const std::function<
+        void(const typename std::vector<Segment>::const_iterator,
+             uint64_t& /* byte_size */,
+             uint64_t& /* docs_count */,
+             uint64_t& /* live_docs_count */)>& getSegmentAttributes,
+      tier::ConsolidationCandidate<Segment>& best) {
+
+        //  sort segments in increasing order of the segment byte size
+        std::sort(sorted_segments.begin(), sorted_segments.end());
+
+        //  We start with a min. window size of 2
+        //  since a window of size 1 will always
+        //  give us a skew of 1.0.
+        uint64_t minWindowSize { tier::ConsolidationConfig::candidate_size };
+        auto front = sorted_segments.begin();
+        auto rear = front + minWindowSize - 1;
+        tier::ConsolidationCandidate<Segment> candidate(front, rear, getSegmentAttributes);
+
+        double prev_score { 1.0 };
+
+        while ((candidate.first() + 1) < sorted_segments.end()) {
+
+          if (!best.initialized || (best.mergeScore > candidate.mergeScore && candidate.mergeBytes <= max_segments_bytes))
```

Comment:
`best` will possibly be initialized with an invalid candidate (that violates the size limit). This will later prevent valid candidates from being selected if they have a worse score.

So I suggest


```suggestion
          if (candidate.mergeBytes <= max_segments_bytes && (!best.initialized || best.mergeScore > candidate.mergeScore))
```

Reply:
And I think the candidates checked here can also be below the min window size, though I haven't checked whether this can cause a problem or not.